Type equivalence support in Crossgen2

davidwrighton opened the pull request:
We currently have a small behavior hole on Windows where if a binary uses TypeEquivalence, it is technically possible to get crossgen2 to generate invalid code.

This fixes that by adding support for type equivalence to the managed type system, and adjusting crossgen2 to use it. This is complicated by the detail that this may generate some type references which cannot be referenced from an R2R file, so additional error checking was needed (this additional error checking is what fixes #67855).

However, while basic support for type equivalence has been added to the type system, support for actually performing type equivalent interface resolution has not been, so devirtualization of type equivalent interface calls is disabled. At this time, I do not expect to see a customer need for implementing that fairly complicated feature

Fixes #67855

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into type_equivalence_crossgen2

davidwrighton committed:
It builds!

davidwrighton committed:
Handle type equivalent method lookup by token

davidwrighton committed:
Virtual overrides with equivalence.

davidwrighton committed:
Enable type equivalent behavior on Windows in Crossgen2

davidwrighton committed:
Fix build

davidwrighton committed:
Add test for some actual signature type equivalence checks, and add handling so that problematic type equivalence cases will cause a runtime jit fallback

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into type_equivalence_crossgen2_modern

davidwrighton committed:
Add type loader validation to type equivalence logic
Cache computation of type equivalence on EcmaType
Fix build, and do first pass of code review

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.TypeEquivalence.cs:
```diff
@@ -0,0 +1,145 @@
+                        return null;
+                    return _data;
+                }
+                lock(this)
```
We don't generally lock on type system objects because with enough unchecked locks we'll eventually get deadlocks (see another spot where I'm policing this: https://github.com/dotnet/runtime/pull/87532#discussion_r1228848088). Would CompareExchange be unacceptable?

davidwrighton replied:
Yes. This was just a bit of laziness on my part.

davidwrighton committed:
Fix code review issue and disable test where it shouldn't work anyways

davidwrighton commented:
@trylek the tests are passing, (except for a couple mono issues that are known). Could you take a look and let me know what you can sign off on, and what I need to find someone else to sign off on?

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs:
```diff
@@ -0,0 +1,312 @@
+                }
+
+                // Generic equivalence only allows the instantiation to be non-equal
+                return thisType.HasSameTypeDefinition(otherType);
```
Perf nit - wouldn't it be faster to first check whether the two types have the same type definition and only then traverse the instantiations?

davidwrighton replied:
Good point. Changed

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs:
```diff
@@ -0,0 +1,312 @@
+                    return false;
+                if (type1.IsInterface)
+                {
+                    if (!type2.IsInterface)
```
Nit - I believe this check is superfluous; you have already made sure at line 132 that <code>type1.IsInterface</code> is the same as <code>type2.IsInterface</code> so it seems to me that you should be able to safely <code>return true</code> in this conditional statement.

davidwrighton replied:
Thank you

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs:
```diff
@@ -0,0 +1,312 @@
+
+                    return true;
+                }
+                else if ((type1.IsEnum != type2.IsEnum) ||  (type1.IsValueType != type2.IsValueType))
```
Nit - it might make the code slightly easier to read if you removed the <code>else</code> statement at line 141 - the previous conditional statements act as "filters" that either directly return the result or proceed to the next statement.

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs:
```diff
@@ -0,0 +1,312 @@
+                    }
+
+                    // Compare the field signatures for equivalence
+                    // TODO: Technically this comparison should include custom modifiers on the field signatures
```
Is this a real possibility that we should track with an issue?

davidwrighton replied:
Eh, no. We don't even check field types on fields correctly today for normal field references, much less custom modifiers.

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs:
```diff
@@ -0,0 +1,312 @@
+                            return field;
+
+                        // Only public instance fields, and literal fields on enums are permitted in type equivalent structures
+                        if (!enumMode || field.IsLiteral)
```
I may be missing something but shouldn't this check be negated? I would assume that <code>!field.IsLiteral</code> is what signifies that the type doesn't support equivalence.

davidwrighton replied:
Yes. Good catch, that inspired me to build a whole test suite, so I found some more issues. Sigh. Thank you though.

trylek commented on src/coreclr/tools/Common/TypeSystem/Common/MethodDesc.cs:
```diff
@@ -284,6 +294,21 @@ private bool Equals(MethodSignature otherSignature, bool allowCovariantReturn)
 
             return false;
+
+            static bool CompareTypeHelper(TypeDesc type1, TypeDesc type2, bool allowEquivalence, StackOverflowProtect visited)
```
Naming nit - both my old C++ background and Compare-like C# APIs typically return plus / minus / zero result of the comparison whereas this feels more like something in the vein of <code>IsTypeEqualHelper</code>. Not a big deal but I had to fight this mental model half a dozen times while reviewing this change.

trylek reviewed (approved):
In general the Crossgen2 logic looks reasonable to me. I'm not a COM expert so that the craziness around GUIDs is somewhat beyond my comprehension, I guess Aaron would be a great pick if you're after more detailed scrutiny in this area. I have shared a couple of observations including one place that I believe to be a bug but I'm not blocking my approval on that as I'm well aware of your long-standing engineering diligence in addressing all review feedback.

davidwrighton committed:
Add test suite for managed type system type equivalence comparisons
- Share test suite with runtime

Fix some issues found in type equivalence comparison found while building said test suite

davidwrighton committed:
Disable type equivalent interfaces and related details from devirtualization in crossgen2

davidwrighton committed:
Disable new test in issues.targets

davidwrighton committed:
Rename helper per code review feedback

davidwrighton committed:
Merge branch 'main' into type_equivalence_crossgen2_modern

trylek commented on src/coreclr/tools/aot/ILCompiler.TypeSystem.Tests/TypeEquivalenceAssembly/TypesWhichDoNotLoad.cs:
```diff
@@ -0,0 +1,68 @@
+        int field;
+    }
+
+    // A case where the type is not both public
```
Nit - I'm missing the purpose of the word *both* in this code comment (unless *both public* is a technical term I haven't yet come across); dtto for the next two test cases.

trylek reviewed (approved):
LGTM, thanks for implementing the extensive test suite!
